torch/csrc/profiler/python/combined_traceback.cpp

```diff
@@ -30,15 +30,24 @@ struct PythonTraceback : public CapturedTraceback::Python {
     {
       std::lock_guard<std::mutex> lock(to_free_frames_mutex);
       for (CapturedTraceback::PyFrame f : to_free_frames) {
-        Py_XDECREF(f.code);
+        Py_XDECREF(f.filename);
+        Py_XDECREF(f.funcname);
       }
       to_free_frames.clear();
     }
     PyFrameObject* f = PyEval_GetFrame();
     Py_XINCREF(f);
     while (f) {
-      frames.emplace_back(
-          CapturedTraceback::PyFrame{PyFrame_GetCode(f), PyFrame_GetLasti(f)});
+      PyCodeObject* code = PyFrame_GetCode(f);
+      Py_INCREF(code->co_filename);
+      Py_INCREF(code->co_name);
+      int lasti = PyFrame_GetLasti(f);
+      frames.emplace_back(CapturedTraceback::PyFrame{
+          code->co_filename,
+          code->co_name,
+          PyCode_Addr2Line(code, lasti),
```

Comment:
Addr2Line is an O(n) search through the bytecodes of the function to find the line number. For similar reasons to why we can't create a weak reference here, we also don't want to do a search for line number for every frame during collection.

Reply:
Sadly, I don't see a way forward with this if we don't calculate the line number here. As we would either need to hold the line number mapping (as you suggested above), which I'm against (explained above too), or hold a borrowed reference to a code object, which might be invalid.

We can completely fix this with [code object watchers added in 3.12](https://docs.python.org/3/c-api/code.html#c.PyCode_AddWatcher) (coincidentally, added by the Cinder team at Meta). However, that won't fix the rest of the versions.

Reply:
I might just copy in the PyCode_Addr2Line source from CPython pre-3.10 and post-3.10. Is there any concern licensing wise doing something like that?

Reply:
This should be ok, there's prior art for this as we often have to copy in source from cpython to access internals that aren't publicly exposed